### DISCUSSION TITLE
Per Route SSR/SPA

### DISCUSSION THREAD

**Discussion by lucaschultz** [nuxt/nuxt#15921] [Questions]
Hey all!

Stumbled upon an [old issue](https://github.com/nuxt/nuxt.js/issues/1468#issuecomment-324692324) where @pi0 explains how to achieve per route SSR using a custom server middleware. Is this still possible using Nuxt 3?

Been trying a (very naive) example without success. Lets assume I wan't every page under `/jobs/` to be SSR and everything else to be an SPA. I set my `nuxt.config.ts` to:

```ts
export default defineNuxtConfig({
  ssr: false
});
```

and created a middleware `server> middleware > ssrJobs.ts` like so:

```ts
export default defineEventHandler((event) => {
  if (event.req.url?.includes("jobs")) { // remember ... naive example
    event.res.ssr = true;
  }
});
```

However, I'm not getting anything pre-rendered from the server. Just what I assume is the entry point for the SPA:

```
❯ curl http://localhost:3000/jobs
<!DOCTYPE html>
<html >

<head >

</head>

<body >
  <div id="__nuxt"></div><script>window.__NUXT__={serverRendered:false,config:{public:{},app:{baseURL:"\u002F",buildAssetsDir:"\u002F_nuxt\u002F",cdnURL:""}}}</script><script type="module" src="/_nuxt/@vite/client"></script><script type="module" src="/_nuxt/entry.mjs"></script>
</body>

</html>
```

What am I doing wrong?

Cheers for your answers 🤘

**Comment by alehano** [comment]
Hello. Same problem here. I have admin section and I'd like to set ssr off for it.
Now I wrapped admin pages with ClientOnly tag. It works ok. But when I reload admin page by browser's F5, I got this ssr related problems in console:

```
runtime-core.esm-bundler.js:38 [Vue warn]: Unhandled error during execution of render function 
  at <ClientOnly>

[Vue warn]: Unhandled error during execution of scheduler flush. This is likely a Vue internals bug. Please open an issue at https://new-issue.vuejs.org/?repo=vuejs/core 
  at <ClientOnly> 

Uncaught (in promise) TypeError: Cannot read properties of null (reading 'Id')
```

The last error happened because on server side Nuxt can't get User.Id due to unauthorized - auth tokens stored in LocalStore and server side can't get access to it.

**Comment by Miles-C** [comment]
I also had this problem, my solution:
```ts
//server/middleware/ssr.ts
export default defineEventHandler((event) => {
  if (event.node.req.url === '/test')
    event.node.req.headers["x-nuxt-no-ssr"] = "true";
});
```

**Comment by TheAlexLichter** [comment]
I highly recommend to use routeRules to solve that problem unless a more "advanced" check is needed ☺️
